test_runner/regress/test_wal_acceptor.py

```diff
@@ -1057,6 +1057,24 @@ def test_restart_endpoint(neon_env_builder: NeonEnvBuilder):
         endpoint.start()
 
 
+# Try restarting endpoint immediately after xlog switch.
+# https://github.com/neondatabase/neon/issues/8911
+def test_restart_endpoint_after_switch_wal(neon_env_builder: NeonEnvBuilder):
+    env = neon_env_builder.init_start()
+
+    endpoint = env.endpoints.create_start("main")
+
+    endpoint.safe_psql("create table t (i int)")
+
+    endpoint.safe_psql("SELECT pg_switch_wal()")
+
+    # we want immediate shutdown to have endpoint restart on xlog switch record,
+    # so prevent shutdown checkpoint.
+    endpoint.stop(mode='immediate')
```

Comment:
as it was originally reported for stop_and_destroy, this is now different (we reuse the directory). should we parameterize the test for all possible variants of stopping?

- immediate + destroy
- graceful + destroy
- immediate + reuse
- graceful + reuse

In general, I don't think we have a lot of tests relying on reusing the endpoint but re-creating it.

Reply:
No, for the test itself directory doesn't matter, what matters is immediate shutdown to prevent any WAL writes after xlog switch, Heikki did stop_and_destroy to trigger it which is the default there. I don't like stop_and_destroy because it destroys datadir along with compute logs, so you can't analyze them. So I always choose to either reuse endpoint or create new one, but stop previous without destroy.